ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,313 @@
+static const uint32_t MinEntriesCapacity = 0x040000; //  4 MB (assuming 16 bytes per VnodeCacheEntry)
+static const uint32_t MaxEntriesCapacity = 0x400000; // 64 MB (assuming 16 bytes per VnodeCacheEntry)
+
+kern_return_t VnodeCache_Init()
+{
+    if (RWLock_IsValid(s_entriesLock))
+    {
+        return KERN_FAILURE;
+    }
+    
+    s_entriesLock = RWLock_Alloc();
+    if (!RWLock_IsValid(s_entriesLock))
+    {
+        return KERN_FAILURE;
+    }
+
+    s_entriesCapacity = Clamp(desiredvnodes * 2u, MinEntriesCapacity, MaxEntriesCapacity);
+    
+    s_entries = Memory_AllocArray<VnodeCacheEntry>(s_entriesCapacity);
+    if (nullptr == s_entries)
+    {
+        s_entriesCapacity = 0;
+        return KERN_RESOURCE_SHORTAGE;
+    }
+    
+    VnodeCache_InvalidateCache(nullptr);
+    
+    PerfTracing_RecordSample(PrjFSPerfCounter_CacheCapacity, 0, s_entriesCapacity);
+    
+    return KERN_SUCCESS;
+}
+
+void VnodeCache_Cleanup()
+{
+    if (nullptr != s_entries)
+    {
+        Memory_FreeArray<VnodeCacheEntry>(s_entries, s_entriesCapacity);
+        s_entries = nullptr;
+        s_entriesCapacity = 0;
+    }
+    
+    if (RWLock_IsValid(s_entriesLock))
+    {
+        RWLock_FreeMemory(&s_entriesLock);
+    }
+}
+
+VirtualizationRootHandle VnodeCache_FindRootForVnode(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheHitCounter,
+    PrjFSPerfCounter cacheMissCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vnode_t _Nonnull vnode,
+    vfs_context_t _Nonnull context,
+    bool invalidateEntry)
+{
+    VirtualizationRootHandle rootHandle = RootHandle_None;
+    uintptr_t vnodeHash = HashVnode(vnode);
+    uint32_t vnodeVid = vnode_vid(vnode);
+    
+    if (!invalidateEntry)
+    {
+        if (TryGetVnodeRootFromCache(vnode, vnodeHash, vnodeVid, rootHandle))
+        {
+            perfTracer->IncrementCount(cacheHitCounter, true /*ignoreSampling*/);
+            return rootHandle;
+        }
+    }
+    
+    perfTracer->IncrementCount(cacheMissCounter, true /*ignoreSampling*/);
+    UpdateCacheForVnode(
+        perfTracer,
+        cacheMissFallbackFunctionCounter,
+        cacheMissFallbackFunctionInnerLoopCounter,
+        context,
+        vnode,
+        vnodeHash,
+        vnodeVid,
+        invalidateEntry,
+        rootHandle);
+    
+    return rootHandle;
+}
+
+void VnodeCache_InvalidateCache(PerfTracer* _Nullable perfTracer)
+{
+    if (nullptr != perfTracer)
+    {
+        perfTracer->IncrementCount(PrjFSPerfCounter_CacheInvalidateCount, true /*ignoreSampling*/);
+    }
+
+    RWLock_AcquireExclusive(s_entriesLock);
+    {
+        InvalidateCache_ExclusiveLocked();
+    }
+    RWLock_ReleaseExclusive(s_entriesLock);
+}
+
+KEXT_STATIC_INLINE void InvalidateCache_ExclusiveLocked()
+{
+    memset(s_entries, 0, s_entriesCapacity * sizeof(VnodeCacheEntry));
+}
+
+KEXT_STATIC_INLINE uintptr_t HashVnode(vnode_t _Nonnull vnode)
```

Comment:
Just double checking HashVnode is the convention here, not something like GetVnodeHashKey

Reply:
Updating to `ComputeVnodeHashKey`